src/Compilers/CSharp/Portable/Symbols/OverriddenOrHiddenMembersHelpers.cs

```diff
@@ -920,43 +905,85 @@ private static int CustomModifierCount(Symbol member)
         /// <summary>
         /// Determine if this method requires a methodimpl table entry to inform the runtime of the override relationship.
         /// </summary>
-        internal static bool RequiresExplicitOverride(this MethodSymbol method)
+        /// <param name="warnAmbiguous">True if we should produce an ambiguity warning per https://github.com/dotnet/roslyn/issues/45453 .</param>
+        internal static bool RequiresExplicitOverride(this MethodSymbol method, out bool warnAmbiguous)
+        {
+            warnAmbiguous = false;
+            if (!method.IsOverride)
+                return false;
+
+            MethodSymbol csharpOverriddenMethod = method.OverriddenMethod;
+            if (csharpOverriddenMethod is null)
+                return false;
+
+            MethodSymbol runtimeOverriddenMethod = method.GetFirstRuntimeOverriddenMethodIgnoringNewSlot(out bool wasAmbiguous);
+            if (csharpOverriddenMethod == runtimeOverriddenMethod && !wasAmbiguous)
+                return false;
+
+            // See https://github.com/dotnet/roslyn/issues/45453. No need to warn when the runtime
+            // supports covariant returns because the methodimpl is unambiguously understood by the runtime.
+            if (method.ContainingAssembly.RuntimeSupportsCovariantReturnsOfClasses)
+                return true;
+
+            // If the method was declared as a covariant return, there will be a compile-time error since the runtime
+            // does not support covariant returns. In this case we do not warn about runtime ambiguity and pretend that
+            // we can use a methodimpl (even though it is of a form not supported by the runtime and would result in a
+            // loader error) so that the symbol APIs produce the most useful result.
+            if (!method.ReturnType.Equals(csharpOverriddenMethod.ReturnType, TypeCompareKind.AllIgnoreOptions))
+                return true;
+
+            // Due to https://github.com/dotnet/runtime/issues/38119 the methodimpl would
+            // appear to the runtime to be ambiguous in some cases.
+            bool methodimplWouldBeAmbiguous = csharpOverriddenMethod.MethodHasRuntimeCollisionInSubstututedForm();
+            if (!methodimplWouldBeAmbiguous)
+                return true;
+
+            // We produce the warning when a methodimpl would be required but would be ambiguous to the runtime
+            warnAmbiguous = true;
+
+            bool overridenMethodContainedInSameTypeAsRuntimeOverriddenMethod =
+                csharpOverriddenMethod.ContainingType.Equals(runtimeOverriddenMethod.ContainingType, TypeCompareKind.CLRSignatureCompareOptions);
```

Comment:
>CLRSignatureCompareOptions [](start = 117, length = 26)

Do we expect custom modifiers and array sizes make a difference for this scenario?

Reply:
Since the same class can only occur once in the inheritance chain, I don't think there can be any differences as a practical matter.

---
In reply to: [448079095](https://github.com/dotnet/roslyn/pull/44025#discussion_r448079095) [](ancestors = 448079095)